openhcl/hcl/src/ioctl/tdx.rs

```diff
@@ -109,6 +111,61 @@ impl ProcessorRunner<'_, Tdx> {
         &self.tdx_vp_context().exit_info
     }
 
+    /// Clears the ipi offload IRR field, returning the previous value.
+    /// This IRR field is used by the kernel to accelerate certain APIC operations.
+    pub fn tdx_vp_ipi_offload_extract_irr(&mut self) -> Option<[u32; 8]> {
+        let ctx = self.tdx_vp_context_mut();
+        // SAFETY: ipi_offload_irr is accessed as an atomic u32 in all
+        // other contexts (kernel code). The layout of an AtomicU32 matches a u32.
+        let offload_irr: &mut [AtomicU32; 8] = unsafe {
+            std::mem::transmute::<&mut [u32; 8], &mut [AtomicU32; 8]>(&mut ctx.ipi_offload_irr)
+        };
+
+        let mut r = [0; 8];
+        let mut found = false;
+        for (irr, r) in offload_irr.iter().zip(r.iter_mut()) {
+            if irr.load(Ordering::Acquire) != 0 {
+                *r = irr.swap(0, Ordering::SeqCst);
+                found = true;
+            }
+        }
+
+        found.then_some(r)
+    }
+
+    /// Sets the kernel IPI offload TMR for the specificed bank.
+    /// This must be kept in sync with the emulated APIC TMR.
+    /// Otherwise, the kernel will not know if an edge-triggered interrupt
+    /// was previously used for a level-triggered interrupt, leading to a
+    /// level-triggered-EOI instead of edge-triggered-EOI.
+    pub fn tdx_vp_ipi_offload_set_tmr(&mut self, bank: usize, val: u32) {
```

Comment:
Since there's no cross-thread use, you might as well just return the array itself, e.g. `pub fn tdx_vp_ipi_offload_tmr(&mut self) -> &mut [u32; 8]`.

Reply:
Changed accordingly in the new version